more/pixels/pixelMintKeeper.js

```diff
@@ -0,0 +1,117 @@
+import { makePrivateName } from '../../util/PrivateName';
+import { insist } from '../../util/insist';
+import { getString } from './types/pixel';
+
+// This custom mintKeeper does the usual recordings of new
+// purses/payments and updated balances but it also allows for a
+// special capability to destroy amounts: effectively remove pixels
+// from the purses or payments they were in. In order to do that, we
+// must continuously record every movement of a pixel to a new purse
+// or payment. We need this functionality in order to have the ability to revoke
+// childPayments/childPurses
+export function makePixelMintKeeper(assay) {
+  // individual pixel to purse/payment
+  const pixelToAsset = new Map();
+
+  // This helper function takes an amount, takes out the pixelList within it,
+  // and makes sure that the mapping of each pixel to asset is updated.
+  function recordPixelsAsAsset(amount, asset) {
+    amount = assay.coerce(amount);
+    const pixelList = assay.quantity(amount);
+    for (const pixel of pixelList) {
+      pixelToAsset.set(getString(pixel), asset);
+    }
+  }
+
+  function makeAssetKeeper() {
+    // asset to amount
+    const assets = makePrivateName();
+    return {
+      // updateAmount and recordNew are the same as the core
+      // mintKeeper, except that we also record the movement of the
+      // pixels when they are called.
+      updateAmount(asset, newAmount) {
+        assets.set(asset, newAmount);
+        recordPixelsAsAsset(newAmount, asset);
+      },
+      recordNew(asset, initialAmount) {
+        assets.init(asset, initialAmount);
+        recordPixelsAsAsset(initialAmount, asset);
+      },
+      getAmount(asset) {
+        return assets.get(asset);
+      },
+      has(asset) {
+        return assets.has(asset);
+      },
+    };
+  }
+
+  const purseKeeper = makeAssetKeeper('purse');
+  const paymentKeeper = makeAssetKeeper('payment');
+
+  // This helper function is used by `destroy` to find the keeper
+  // associated with the asset, when we retrieved the asset from the
+  // `pixelToAsset` map and we don't yet know whether the asset
+  // is a purse or payment
+  function getKeeper(asset) {
+    if (purseKeeper.has(asset)) {
+      return purseKeeper;
+    }
+    if (paymentKeeper.has(asset)) {
+      return paymentKeeper;
+    }
+    throw new Error(
+      `asset ${asset.getName()} was not recognized as a purse or a payment`,
+    );
+  }
+
+  const pixelMintKeeper = {
```

Comment:
Please re-review everything for missing calls to harden. By habit, I would have written this

```js
const pixelMintKeeper = harden({
```

If not hardened here, it should at least be hardened before being returned. We won't get all of these reliably until we have a static checker for these, but we need to develop the habits of writing code that will be closer to passing such a static checker. Each one we miss is likely a security hole.

I know I have not been vigilant about this in reviews in general. We all must be.

Reply:
I've looked for the creation of object literals and added `harden` where it was lacking